File: TvNavigationLayout-java/app/src/main/java/com/hxxdashai/android/navigation/widget/NavigationCursorView.java
package com.hxxdashai.android.navigation.widget;

import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ImageView;

/**
 * Created by Mr.T on 2018/4/25.
 */

public class NavigationCursorView extends ImageView {

    private long mDuration = 0L;
    private int mLastLocation = 0;
    private final AnimatorSet set = new AnimatorSet();

    public NavigationCursorView(Context context) {
        this(context, null);
    }

    public NavigationCursorView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NavigationCursorView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setVisibility(View.INVISIBLE);
    }

    public final void fsatJumpTo(int location) {
        mDuration = 1L;
        jumpTo(location);
        mDuration = 200L;
    }

    public final void jumpTo(int location) {
        int realLocation = location - getWidth() / 2;
        if (mLastLocation == realLocation) return;
        if (set.isRunning()) set.cancel();
        createAnimator(realLocation).start();
        mLastLocation = realLocation;
    }

    public final AnimatorSet createAnimator(int location) {
        ObjectAnimator translationX = ObjectAnimator.ofFloat(this, "translationX", (float) mLastLocation, (float) location);
        ObjectAnimator rotationY = ObjectAnimator.ofFloat(this, "rotationY", 0.0F, mLastLocation > location ? -180.0F : 180.0F, 0.0F);
        ObjectAnimator scaleX = ObjectAnimator.ofFloat(this, "scaleX", 1.0F, 0.2F, 1.0F);
        set.setDuration(mDuration);
        set.playTogether(translationX, rotationY, scaleX);
        return set;
    }
}

File: TvNavigationLayout-kotlin/app/src/main/java/com/hxxdashai/android/navigation/widget/NavigationCursorView.kt
package com.hxxdashai.android.navigation.widget

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ImageView

/**
 * Created by Mr.T on 2018/3/29.
 */
class NavigationCursorView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : ImageView(context, attrs, defStyleAttr) {

    init {
        visibility = View.INVISIBLE
    }

    private var mDuration = 0L
    private var mLastLocation = 0
    private val set = AnimatorSet()

    /**
     * 光标切换位置时候调用(忽略动画)初始化时候调用
     */
    fun fsatJumpTo(location: Int) {
        mDuration = 1L
        jumpTo(location)
        mDuration = 200L
    }

    /**
     * 光标切换位置时候调用(有动画)
     */
    fun jumpTo(location: Int) {
        val realLocation = location - width / 2
        if (mLastLocation == realLocation) return
        if (set.isRunning) set.cancel()
        createAnimator(realLocation).start()
        mLastLocation = realLocation
    }

    fun createAnimator(location: Int): AnimatorSet {
        val translationX = ObjectAnimator.ofFloat(this, "translationX", mLastLocation.toFloat(), location.toFloat())
        val rotationY = ObjectAnimator.ofFloat(this, "rotationY", 0.0f, if (mLastLocation > location) -180.0f else 180.0f, 0.0f)
        val scaleX = ObjectAnimator.ofFloat(this, "scaleX", 1f, 0.2f, 1f)
        set.duration = mDuration
        set.playTogether(translationX, rotationY, scaleX)
        return set
    }
}